File: marketdna-web/src/types/regime.ts
export interface RegimeComponents {
  price_score: number      // 0–40
  alignment_score: number  // 0–30
  slope_score: number      // 0–30
}

export interface SMASummary {
  sma20: number
  sma50: number
  sma100: number
  sma200: number
  above_sma20: boolean
  above_sma50: boolean
  above_sma100: boolean
  above_sma200: boolean
}

export interface SlopeInfo {
  sma20_slope: number
  sma50_slope: number
  sma200_slope: number
  sma20_rising: boolean
  sma50_rising: boolean
  sma200_rising: boolean
}

export interface RegimeHistoryPoint {
  date: string
  close: number
  regime_score: number
}

export interface RegimeScoreResult {
  symbol: string
  date: string
  close: number
  regime_score: number
  regime_label: string
  components: RegimeComponents
  sma: SMASummary
  slope: SlopeInfo
  sma_fully_aligned: boolean
  history: RegimeHistoryPoint[]
  computed_at: string
}

export interface BreadthHistoryPoint {
  date: string
  breadth_score: number
  pct_above_sma20: number
  pct_above_sma50: number
  pct_above_sma200: number
}

export interface BreadthScoreResult {
  date: string
  pct_above_sma20: number
  pct_above_sma50: number
  pct_above_sma200: number
  count_above_sma20: number
  count_above_sma50: number
  count_above_sma200: number
  total_symbols: number
  breadth_score: number
  breadth_label: string
  history: BreadthHistoryPoint[]
  computed_at: string
}

export interface StockRegimeSummary {
  symbol: string
  regime_score: number
  regime_label: string
  close: number
  above_sma200: boolean
  sma_fully_aligned: boolean
  price_score: number
  alignment_score: number
  slope_score: number
}

export interface MarketRegimeSnapshot {
  date: string
  breadth: BreadthScoreResult
  stocks: StockRegimeSummary[]
  avg_regime_score: number
  strong_bull_count: number
  bull_count: number
  neutral_count: number
  bear_count: number
  strong_bear_count: number
  computed_at: string
}

// ─── Display helpers ──────────────────────────────────────────────────────────

export const REGIME_COLOR: Record<string, string> = {
  'Strong Bull':   '#22c55e',
  'Moderate Bull': '#86efac',
  'Neutral':       '#fbbf24',
  'Moderate Bear': '#f87171',
  'Strong Bear':   '#ef4444',
}

export const REGIME_BG: Record<string, string> = {
  'Strong Bull':   'rgba(34,197,94,0.08)',
  'Moderate Bull': 'rgba(134,239,172,0.08)',
  'Neutral':       'rgba(251,191,36,0.08)',
  'Moderate Bear': 'rgba(248,113,113,0.08)',
  'Strong Bear':   'rgba(239,68,68,0.08)',
}

export function regimeColor(score: number): string {
  if (score >= 80) return '#22c55e'
  if (score >= 60) return '#86efac'
  if (score >= 40) return '#fbbf24'
  if (score >= 20) return '#f87171'
  return '#ef4444'
}

File: marketdna-backend/app/models/regime.py
"""Pydantic models for Regime Score and Market Breadth Score."""
from pydantic import BaseModel


class RegimeComponents(BaseModel):
    """Breakdown of the three sub-scores that sum to the regime score."""
    price_score: float     # 0–40  — price vs SMA position
    alignment_score: float # 0–30  — SMA bull-stack alignment
    slope_score: float     # 0–30  — SMA trend direction


class SMASummary(BaseModel):
    """Current SMA levels and whether price is above each one."""
    sma20: float
    sma50: float
    sma100: float
    sma200: float
    above_sma20: bool
    above_sma50: bool
    above_sma100: bool
    above_sma200: bool


class SlopeInfo(BaseModel):
    """SMA slope values (% change over lookback) and direction flags."""
    sma20_slope: float   # % change over 5 trading days
    sma50_slope: float   # % change over 10 trading days
    sma200_slope: float  # % change over 20 trading days
    sma20_rising: bool
    sma50_rising: bool
    sma200_rising: bool


class RegimeHistoryPoint(BaseModel):
    date: str
    close: float
    regime_score: float


class RegimeScoreResult(BaseModel):
    symbol: str
    date: str
    close: float
    regime_score: float      # 0–100 composite
    regime_label: str        # "Strong Bull" | "Moderate Bull" | "Neutral" | "Moderate Bear" | "Strong Bear"
    components: RegimeComponents
    sma: SMASummary
    slope: SlopeInfo
    sma_fully_aligned: bool  # True when SMA20 > SMA50 > SMA100 > SMA200
    history: list[RegimeHistoryPoint]  # last 252 trading days
    computed_at: str


class BreadthHistoryPoint(BaseModel):
    date: str
    breadth_score: float
    pct_above_sma20: float
    pct_above_sma50: float
    pct_above_sma200: float


class BreadthScoreResult(BaseModel):
    date: str
    pct_above_sma20: float   # % of NIFTY 50 stocks with close > SMA20
    pct_above_sma50: float
    pct_above_sma200: float
    count_above_sma20: int
    count_above_sma50: int
    count_above_sma200: int
    total_symbols: int
    breadth_score: float     # 0–100 weighted composite
    breadth_label: str       # "Broad Participation" | "Moderate Breadth" | "Narrow Breadth" | "Poor Breadth"
    history: list[BreadthHistoryPoint]  # last 252 trading days
    computed_at: str


class StockRegimeSummary(BaseModel):
    symbol: str
    regime_score: float
    regime_label: str
    close: float
    above_sma200: bool
    sma_fully_aligned: bool
    price_score: float
    alignment_score: float
    slope_score: float


class MarketRegimeSnapshot(BaseModel):
    date: str
    breadth: BreadthScoreResult
    stocks: list[StockRegimeSummary]
    avg_regime_score: float
    strong_bull_count: int   # regime_score >= 80
    bull_count: int          # 60 ≤ score < 80
    neutral_count: int       # 40 ≤ score < 60
    bear_count: int          # 20 ≤ score < 40
    strong_bear_count: int   # score < 20
    computed_at: str
